File: pkg/dashboard/panel.go
package dashboard

import (
	"strings"

	"github.com/mattn/go-runewidth"
)

type panelRow struct {
	key, val string
	raw      bool
}

type Panel struct {
	title        string
	width        int
	height       int
	padding      int
	scrollOffset int
	rows         []panelRow
}

func NewPanel(title string, width, height int) *Panel {
	return &Panel{title: title, width: width, height: height}
}

func (p *Panel) SetPadding(n int)      { p.padding = n }
func (p *Panel) SetScrollOffset(n int) { p.scrollOffset = n }

func (p *Panel) AddKV(key, val string) {
	p.rows = append(p.rows, panelRow{key: key, val: val})
}

func (p *Panel) AddRaw(text string) {
	p.rows = append(p.rows, panelRow{val: text, raw: true})
}

func (p *Panel) Render() []string {
	inner := p.width - 2

	titleText := " " + p.title + " "
	barLen := inner - runewidth.StringWidth(titleText)
	left := barLen / 2
	right := barLen - left
	top := "┌" + strings.Repeat("─", left) + titleText + strings.Repeat("─", right) + "┐"
	bottom := "└" + strings.Repeat("─", inner) + "┘"

	allContent := p.renderRows(inner)
	contentH := p.height - 2
	needScroll := len(allContent) > contentH

	visible := allContent
	if needScroll {
		end := p.scrollOffset + contentH
		if end > len(allContent) {
			end = len(allContent)
		}
		visible = allContent[p.scrollOffset:end]
	}

	empty := "│" + strings.Repeat(" ", inner) + "│"
	for len(visible) < contentH {
		visible = append(visible, empty)
	}

	if needScroll {
		visible = applyScrollbar(visible, p.scrollOffset, len(allContent), contentH)
	}

	lines := make([]string, 0, p.height)
	lines = append(lines, top)
	lines = append(lines, visible...)
	lines = append(lines, bottom)
	return lines
}

func (p *Panel) renderRows(inner int) []string {
	pad := p.padding
	contentW := inner - pad*2

	maxKey := 0
	for _, r := range p.rows {
		if !r.raw && runewidth.StringWidth(r.key) > maxKey {
			maxKey = runewidth.StringWidth(r.key)
		}
	}

	var lines []string
	for _, r := range p.rows {
		var text string
		if r.raw {
			text = truncatePad(r.val, contentW)
		} else {
			prefix := " " + padRight(r.key, maxKey) + ": "
			remaining := contentW - runewidth.StringWidth(prefix)
			text = prefix + truncatePad(r.val, remaining)
		}
		padStr := strings.Repeat(" ", pad)
		lines = append(lines, "│"+padStr+text+padStr+"│")
	}
	return lines
}

func applyScrollbar(lines []string, offset, total, trackH int) []string {
	if total <= trackH {
		return lines
	}
	thumbSize := max(1, trackH*trackH/total)
	thumbPos := trackH * offset / total
	thumbPos = min(thumbPos, trackH-thumbSize)

	result := make([]string, len(lines))
	for i, line := range lines {
		ch := "│"
		if i >= thumbPos && i < thumbPos+thumbSize {
			ch = "┃"
		}
		runes := []rune(line)
		if len(runes) > 0 {
			runes[len(runes)-1] = []rune(ch)[0]
		}
		result[i] = string(runes)
	}
	return result
}

func truncatePad(s string, width int) string {
	s = truncate(s, width)
	return padRight(s, width)
}

func truncate(s string, width int) string {
	w := 0
	for i, ch := range s {
		cw := runewidth.RuneWidth(ch)
		if w+cw > width {
			return s[:i] + strings.Repeat(" ", width-w)
		}
		w += cw
	}
	return s
}

func padRight(s string, width int) string {
	w := runewidth.StringWidth(s)
	if w >= width {
		return s
	}
	return s + strings.Repeat(" ", width-w)
}

File: tl_mr6400/panel.py
import unicodedata


def display_width(text: str) -> int:
    w = 0
    for ch in text:
        eaw = unicodedata.east_asian_width(ch)
        w += 2 if eaw in ("W", "F") else 1
    return w


def _pad(text: str, width: int) -> str:
    return text + " " * (width - display_width(text))


def _truncate(text: str, width: int) -> str:
    w = 0
    for i, ch in enumerate(text):
        cw = 2 if unicodedata.east_asian_width(ch) in ("W", "F") else 1
        if w + cw > width:
            return text[:i] + " " * (width - w)
        w += cw
    return text


class Panel:
    def __init__(self, title: str, width: int, height: int):
        self.title = title
        self.width = width
        self.height = height
        self._rows: list[tuple[str, str] | str] = []
        self._styles: list[str] = []

    def add(self, key: str, value: str, style: str = "key"):
        self._rows.append((key, value))
        self._styles.append(style)

    def add_raw(self, text: str, style: str = "value"):
        self._rows.append(text)
        self._styles.append(style)

    def render(self) -> list[tuple[str, str]]:
        w = self.width
        inner = w - 2

        title_text = f" {self.title} "
        bar_len = inner - len(title_text)
        left_bar = bar_len // 2
        right_bar = bar_len - left_bar
        top = f"┌{'─' * left_bar}{title_text}{'─' * right_bar}┐"
        bottom = f"└{'─' * inner}┘"

        content_lines = self._render_rows(inner)

        content_height = self.height - 2
        empty_line = f"│{' ' * inner}│"
        while len(content_lines) < content_height:
            content_lines.append((empty_line, "empty"))
        content_lines = content_lines[:content_height]

        result = [(top, "title")]
        result.extend(content_lines)
        result.append((bottom, "border"))
        return result

    def _render_rows(self, inner: int) -> list[tuple[str, str]]:
        max_key = max(
            (len(row[0]) for row in self._rows if isinstance(row, tuple)),
            default=0,
        )

        lines = []
        for i, row in enumerate(self._rows):
            style = self._styles[i] if i < len(self._styles) else "value"
            if isinstance(row, str):
                text = _truncate(row, inner)
                text = _pad(text, inner)
            else:
                key, val = row
                prefix = f" {key.ljust(max_key)}: "
                remaining = inner - len(prefix)
                val_text = _truncate(val, remaining)
                val_text = _pad(val_text, remaining)
                text = prefix + val_text
            lines.append((f"│{text}│", style))
        return lines


class PanelGrid:
    @staticmethod
    def horizontal(panels: list['Panel']) -> list[tuple[str, str]]:
        rendered = [p.render() for p in panels]
        max_h = max(len(r) for r in rendered)

        for i, r in enumerate(rendered):
            w = panels[i].width
            empty = f"│{' ' * (w - 2)}│"
            while len(r) < max_h:
                r.append((empty, "empty"))

        lines = []
        for row_idx in range(max_h):
            texts = [r[row_idx][0] for r in rendered]
            style = rendered[0][row_idx][1]
            lines.append(("".join(texts), style))
        return lines

    @staticmethod
    def vertical(panels: list['Panel']) -> list[tuple[str, str]]:
        lines = []
        for p in panels:
            lines.extend(p.render())
        return lines
